src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+    ///         - Has no overloads but its current signature receives an optional ct=default, currently implicit, or...
+    ///         - Has a method overload with the exact same arguments in the same order, plus one ct parameter at the end.
+    ///         
+    /// Conditions for negative cases:
+    ///     - The containing method signature does not receive a ct parameter.
+    ///     - The invocation method signature receives a ct and one is already being explicitly passed, or...
+    ///     - The invocation method does not have an overload with the exact same arguments that also receives a ct, or...
+    ///     - The invocation method only has overloads that receive more than one ct.
+    /// </summary>
+    public abstract class ForwardCancellationTokenToInvocationsAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2016";
+
+        protected abstract SyntaxNode? GetMethodNameNode(SyntaxNode invocationNode);
+
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToInvocationsDescription),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
+
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToInvocationsMessage),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToInvocationsTitle),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
+        internal static DiagnosticDescriptor ForwardCancellationTokenToInvocationsRule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(ForwardCancellationTokenToInvocationsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
```

Comment:
```suggestion
        private static void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
```

Reply:
I can't make the change because I get the warning:

> RS1012: 'context' does not register any analyzer actions. Consider moving actions registered in 'Initialize' that depend on this start action to 'context'.

Reply:
It's because I would also have to make the method `ShouldDiagnose` static so that the analyzer action gets registered, and I can't make that class static because it has abstract methods and those can't be made static:

> CS0112: A static member 'ForwardCancellationTokenToInvocationsAnalyzer.ArgumentsImplicitOrNamed(INamedTypeSymbol, ImmutableArray<IArgumentOperation>)' cannot be marked as override, virtual, or abstract